### PR TITLE
Upgrade `arrow` to `16`

### PR DESCRIPTION
**Pull request by galipremsagar**
## Description
This PR upgrades `arrow` to `16`. This PR also contains fixes to pytests because of breaking API changes in pyarrow.

## Checklist
- [x] I am familiar with the [Contributing Guidelines](https://github.com/rapidsai/cudf/blob/HEAD/CONTRIBUTING.md).
- [x] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.

### CODE REVIEW

File: dependencies.yaml [hunk 1]
```diff
@@ -312,27 +312,27 @@ dependencies:
...
+          - libarrow-acero==16.0.0.*
+          - libarrow-dataset==16.0.0.*
+          - libarrow==16.0.0.*
+          - libparquet==16.0.0.*
   libarrow_run:
     common:
       - output_types: conda
         packages:
           # Allow runtime version to float up to minor version
           # Disallow libarrow 14.0.0 due to a CVE
```
Comment (bdice) [inline dependencies.yaml:324]:
```suggestion
```

File: dependencies.yaml [hunk 2]
```diff
@@ -312,27 +312,27 @@ dependencies:
...
+          - libarrow-acero>=16.0.0,<17.0.0a0
+          - libarrow-dataset>=16.0.0,<17.0.0a0
+          - libarrow>=16.0.0,<17.0.0a0
+          - libparquet>=16.0.0,<17.0.0a0
   pyarrow_run:
     common:
       - output_types: [conda, requirements, pyproject]
         packages:
           # Allow runtime version to float up to minor version
           # Disallow pyarrow 14.0.0 due to a CVE
```
Comment (bdice) [inline dependencies.yaml:334]:
```suggestion
```

**Review by bdice — APPROVED** [review]

File: dependencies.yaml (hunk 1 above)
Comment (galipremsagar) [reply dependencies.yaml]: Ugh! My bad. Thanks for catching this one.

File: python/cudf/cudf/io/parquet.py [hunk 3]
```diff
@@ -993,15 +993,10 @@ def to_parquet(
...
 
-        # Convert partition_file_name to a call back
-        if partition_file_name:
-            partition_file_name = lambda x: partition_file_name  # noqa: E731
-
         pa_table = df.to_arrow(preserve_index=index)
         return pq.write_to_dataset(
             pa_table,
             root_path=path,
-            partition_filename_cb=partition_file_name,
```
Comment (rjzamora) [inline python/cudf/cudf/io/parquet.py:1004]: Do we need to warn or raise if the user specifies `partition_file_name` and the engine is not "cudf"?

**Review by rjzamora — APPROVED** [review]
One question/comment. Otherwise LGTM - Thanks!

File: python/cudf/cudf/io/parquet.py (hunk 3 above)
Comment (galipremsagar) [reply python/cudf/cudf/io/parquet.py:1004]: We might get very noisy because we have many such parameters, I documented the change.

**Pull request merged**